doc/design/model_zoo.md

```diff
@@ -0,0 +1,104 @@
+
+Such design flaw roots from the ignoring of the fact that model definitions are code that has versions. Another important aspect is that we often build the source code into some release-ready form.  Once we noticed these facts, we can use version management tools and Git and release engineering tools like Docker.  Here follows our proposal.
+
+1. A collection of model definitions is a Git repository of source files. 
+1. To describe dependencies, we require a Dockerfile at the root directory of the repository.
+1. To release a repository, we checkout the specific version and run `docker run` with the Dockerfile.
+1. Each release has a globally unique ID (GUID) represented by the hash ID of the Docker image.
+1. The Docker image must support a protocol of command-line arguments so that SQLFlow can pass in required information like hyperparameters from the `docker run` command.
+
+### Argument Protocol
+
+For a SELECT program using the SQLFlow syntax extension, the SQLFlow server converts it into a *submitter* program, usually in Python.
+
+Consider the training statement at the beginning of this document. Suppose that the Python class `MyDNNRegressor` is in source file `my_dnn_regressor.py` in the repository https://github.com/a_data_scientist/regressors, which builds into Docker image `a_data_scientist/regressors`.  Let us also assume that the submitter program is `/var/sqlflow/submitters/an_analyst/my_first_model.py`.  SQLFlow can run the submitter using the Docker image as follows:
+
+```bash
+docker run --rm -it \
+  -v /var/sqlflow/submitters:/submitters \
+  a_data_scientist/regressors \
+    python /submitters/an_analyst/my_first_model.py
+```
+
+If we run SQLFlow in a Docker container, to allow it to run another Docker container, we must enable the Docker-in-Docker feature, say, following [this blog post](https://itnext.io/docker-in-docker-521958d34efd).  Suppose that the SQLFlow server runs in a Kubernetes cluster, where we cannot enable Docker-in-Docker. In such a case, we can make the SQLFlow server call Kubernetes API to run a Pod that executes the above `docker run` command line.
+
+ There could be more than one model definition repositories and Docker images. We need to be able to choose the right one that contains the model `MyDNNRegressor`.  We might want to require that the Docker image, when passed the specific argument `list`, prints the model definition names it contains.  For example, the following command prints a name in the form of `a_data_scientist.MyDNNRegressor`.
+
+```bash
+docker run --rm -it a_data_scientist/regressors list
+```
+
+The submitter program `my_first_model.py`, running in the model definition Docker container, should be able to submit a (distributed) training job to a preconfigured cluster.  Once the job completes, the submitter program adds, or edits, a row in a **model zoo table**.
+
```

Comment:
Use a database to store the saved model is:

1. hard to publish or share the weights to other users.
1. users are using different databases, it can not be commonly used for all open source users.
1. for some database engines it's not suitable to store model parameters in the database like hive. Most of the saved models are of "small size" for HDFS

Reply:
I think the model zoo database is part of the SQLFlow server. It could be embedding DBs like SQLite.